### PR TITLE
fix hotchemi/PermissionsDispatcher#51 based on @aurae sketch code

### PR DESCRIPTION
**Pull request by sibeliusseraphini**

### CODE REVIEW

**Comment by aurae** [comment]
Lgtm, aside from the hard-coded values that I included in the code, not remembering that constants get resolved to Strings anyways.

**Comment by sibeliusseraphini** [comment]
@aurae I've tried using only the put, but didn't work either, only using the hardcoded strings

**Comment by hotchemi** [comment]
Could you add the test code?

File: library/src/main/android/permissions/dispatcher/PermissionUtils.java [hunk 1]
```diff
@@ -9,13 +10,30 @@
...
+import static android.os.Build.VERSION_CODES.KITKAT_WATCH;
+import static android.os.Build.VERSION_CODES.JELLY_BEAN;
 import static android.support.v4.content.PermissionChecker.checkSelfPermission;
 
 public final class PermissionUtils {
 
+    // Map of dangerous permissions introduced in later framework versions.
+    // Used to conditionally bypass permission-hold checks on older devices.
+    private static final SimpleArrayMap<String, Integer> MIN_SDK_PERMISSIONS;
+    static {
```
Comment (hotchemi) [inline library/src/main/android/permissions/dispatcher/PermissionUtils.java]: How did you find these candidates? by hand?

Comment (aurae) [inline library/src/main/android/permissions/dispatcher/PermissionUtils.java]: Yes, I hand-picked these permissions using the Android docs for Manifest.permission](http://developer.android.com/reference/android/Manifest.permission.html), extracting all permissions that were listed as "dangerous" **and** added in a later API version.

**Comment by aurae** [comment]
When using the constants, what didn't work for you? I'll verify this later today, but there shouldn't be any difference in behaviour, since those values get compiled into their String contents anyways.

**Comment by sibeliusseraphini** [comment]
My device (android 4.0) didn't recognize Manifest.permission.READ_EXTERNAL_STORAGE, on monday I will try to add some tests using roboeletric to test it against differents versions of Android devices

**Comment by aurae** [comment]
It would make sense for your 4.0 device to disregard `READ_EXTERNAL_STORAGE`, since that wasn't added to the platform until 4.1.

**Comment by sibeliusseraphini** [comment]
If I ask READ_EXTERNAL_STORAGE permissions using Android M style, it will always denied it for my older device, when I use the hardcoded strings it works, but with Manifest.permission.READ_EXTERNAL_STORAGE it doesn't work

**Comment by aurae** [comment]
I see... this is very odd indeed. I try to find time today to check this out myself. Thanks for your report!

**Comment by aurae** [comment]
Okay, I just double-checked the code, and the usage of `Manifest.permission.*` constants is working as expected. There must've been something off about your code, since I am able to verify that the behaviour of API-level dependent permissions works without the hard-coded strings as well.

Using the extended implementation of `hasSelfPermissions()` and `MIN_SDK_PERMISSIONS` populated as such:
```java
private static final SimpleArrayMap<String, Integer> MIN_SDK_PERMISSIONS;
static {
	MIN_SDK_PERMISSIONS = new SimpleArrayMap<>(6);
	MIN_SDK_PERMISSIONS.put(Manifest.permission.ADD_VOICEMAIL, ICE_CREAM_SANDWICH);
	MIN_SDK_PERMISSIONS.put(Manifest.permission.BODY_SENSORS, KITKAT_WATCH);
	MIN_SDK_PERMISSIONS.put(Manifest.permission.READ_CALL_LOG, JELLY_BEAN);
	MIN_SDK_PERMISSIONS.put(Manifest.permission.READ_EXTERNAL_STORAGE, JELLY_BEAN);
	MIN_SDK_PERMISSIONS.put(Manifest.permission.USE_SIP, GINGERBREAD);
	MIN_SDK_PERMISSIONS.put(Manifest.permission.WRITE_CALL_LOG, JELLY_BEAN);
}
```

Given the following statement (in an Activity's `onCreate()`):
```java
Log.d("Check", "Permission Value: " + Manifest.permission.READ_EXTERNAL_STORAGE + "; do I have that? " + hasSelfPermissions(this, Manifest.permission.READ_EXTERNAL_STORAGE));
```

This prints the following on a device running 2.3.7 Gingerbread:

```
12-14 08:37:40.364 7410-7410/com.example D/Check: Permission Value: android.permission.READ_EXTERNAL_STORAGE; do I have that? true
```

Therefore, please update your PR to use the constants mentioned above, so that we can merge it into the codebase.

**Comment by sibeliusseraphini** [comment]
Sorry, it was my fault, now it is working fine, thanks.

**Comment by aurae** [comment]
Thanks!

**Comment by hotchemi** [comment]
@sibeliusseraphini @aurae Could you add unit test?

**Comment by hotchemi** [comment]
I think we should not merge pr without unit test...

**Comment by aurae** [comment]
Efficiently unit testing API-level dependent code requires us to set up a whole new test environment, using Mockito for instance. I didn't find it *that* necessary to do for this case right away, since it's a fairly simple concept, however I do see your point. I probably can't get to this until after Christmas, though.

**Comment by sibeliusseraphini** [comment]
I think we should run this test using [Instrumented Unit Tests](http://developer.android.com/training/testing/unit-testing/instrumented-unit-tests.html), because the tests should run on real devices with different API levels

**Comment by aurae** [comment]
I have [added some unit tests for this](https://github.com/hotchemi/PermissionsDispatcher/blob/master/library/src/test/java/permissions/dispatcher/ApiLevelTestSuite.java).

**Pull request merged**